src/main/java/com/google/api/codegen/util/js/JSCommentReformatter.java

```diff
@@ -14,50 +14,22 @@
  */
 package com.google.api.codegen.util.js;
 
-import com.google.api.codegen.CommentPatterns;
 import com.google.api.codegen.util.CommentReformatter;
+import com.google.api.codegen.util.CommentTransformer;
+import com.google.api.codegen.util.ProtoLinkPattern;
 import com.google.api.tools.framework.model.ProtoElement;
 import com.google.api.tools.framework.model.ProtoFile;
 import com.google.common.collect.ImmutableSet;
-import java.util.regex.Matcher;
 
 public class JSCommentReformatter implements CommentReformatter {
+
   @Override
   public String reformat(String comment) {
-    comment = reformatProtoMarkdownLinks(comment);
-    comment = reformatCloudMarkdownLinks(comment);
-    return comment.trim();
-  }
-
-  /** Returns a string with all proto markdown links formatted to JSDoc style. */
-  private String reformatProtoMarkdownLinks(String comment) {
-    StringBuffer sb = new StringBuffer();
-    Matcher m = CommentPatterns.PROTO_LINK_PATTERN.matcher(comment);
-    if (!m.find()) {
-      return comment;
-    }
-    do {
-      // proto display name may contain '$' which needs to be escaped using Matcher.quoteReplacement
-      m.appendReplacement(sb, Matcher.quoteReplacement(String.format("{@link %s}", m.group(1))));
-    } while (m.find());
-    m.appendTail(sb);
-    return sb.toString();
-  }
-
-  /** Returns a string with all cloud markdown links formatted to JSDoc style. */
-  private String reformatCloudMarkdownLinks(String comment) {
-    StringBuffer sb = new StringBuffer();
-    Matcher m = CommentPatterns.CLOUD_LINK_PATTERN.matcher(comment);
-    if (!m.find()) {
-      return comment;
-    }
-    do {
-      String url = "https://cloud.google.com" + m.group(2);
-      // cloud markdown links may contain '$' which needs to be escaped using Matcher.quoteReplacement
-      m.appendReplacement(sb, Matcher.quoteReplacement(String.format("[%s](%s)", m.group(1), url)));
-    } while (m.find());
-    m.appendTail(sb);
-    return sb.toString();
+    return CommentTransformer.of(comment)
+        .transform(ProtoLinkPattern.PROTO.createTransformation("{@link %s}"))
+        .transform(ProtoLinkPattern.CLOUD.createTransformation("[%s](%s)"))
```

Comment:
I think the following might be easier to understand:

```
  .transform(PROTO_ELEM_REF.toFormat("{@link $TYPE_REF}"))
  .transform(RELATIVE_URL.toDomain("https://cloud.google.com").toFormat("[$TITLE]($URL)"))
```

Points:
* I don't think that a pattern matcher for relative URLs should know inherently that the URLs are for Cloud - I think that should be informed by the context (Think of non-Cloud APIs)
* I think we should do something that at least looks like named variables; this makes the transformations a lot clearer. Also, it could help remove `Matcher` from the interface of everything.
* The names of the pattern matchers should be more explicit